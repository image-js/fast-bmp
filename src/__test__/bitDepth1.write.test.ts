import { encode, decode } from '..';

import { testHeaderDataEncode } from './testHeaderDataEncode';

const data = {
  width: 0,
  height: 0,
  data: new Uint8Array(),
  bitDepth: 1,
  components: 1,
  channels: 1,
};

describe('encode image with bitDepth of 1', () => {
  it('encode a 5x5 image', () => {
    // 0 0 0 0 0
    // 0 1 1 1 0
    // 0 1 0 1 0
    // 0 1 1 1 0
    // 0 0 0 0 0
    data.width = 5;
    data.height = 5;
    data.data = new Uint8Array([
      0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
      0,
    ]);
    testHeaderDataEncode(data, '5x5.bmp');
    expect(encode(data)).toMatchSnapshot();
  });

  it('encode a 1x5 image', () => {
    // 0
    // 1
    // 0
    // 1
    // 1
    data.width = 1;
    data.height = 5;
    data.data = new Uint8Array([0, 1, 0, 1, 1]);
    testHeaderDataEncode(data, '1x5.bmp');
    expect(encode(data)).toMatchSnapshot();
  });

  it('encode a 5x1 image', () => {
    // 1 0 1 0 0
    data.width = 5;
    data.height = 1;
    data.data = new Uint8Array([1, 0, 1, 0, 0]);
    testHeaderDataEncode(data, '5x1.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });

  it('encode a 6x4 image', () => {
    data.width = 6;
    data.height = 4;
    data.data = new Uint8Array([
      1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
    ]);
    testHeaderDataEncode(data, '6x4.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });

  it('encode a 62x4', () => {
    data.width = 62;
    data.height = 4;
    data.data = new Uint8Array([
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    ]);
    testHeaderDataEncode(data, '62x4.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });

  it('encode a 10x2 image', () => {
    // 1 1 1 0 0 1 0 1 0 1
    // 1 0 1 0 1 0 0 1 1 1
    data.width = 10;
    data.height = 2;
    data.data = new Uint8Array([
      1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0,
    ]);
    testHeaderDataEncode(data, '10x2.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });

  it('encode image with exactly 4 bytes width', () => {
    data.width = 32;
    data.height = 2;
    data.data = new Uint8Array([
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    ]);
    testHeaderDataEncode(data, '32x2.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });

  it('encode image with more that 4 bytes width', () => {
    data.width = 42;
    data.height = 2;
    data.data = new Uint8Array([
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    ]);
    testHeaderDataEncode(data, '42x2.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });
  it('encode image where skipBit can equal relOffset on the last column', () => {
    data.width = 60;
    data.height = 4;
    data.data = new Uint8Array([
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    ]);
    testHeaderDataEncode(data, '60x4.bmp');
    expect(encode(data).slice(146)).toMatchSnapshot();
  });
  it('encode image after decode must give the same image', () => {
    data.width = 60;
    data.height = 4;
    data.data = new Uint8Array([
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
    ]);
    const image = encode(data);

    const result = decode(Buffer.from(image));
    expect(result).toEqual(data);
  });
});
