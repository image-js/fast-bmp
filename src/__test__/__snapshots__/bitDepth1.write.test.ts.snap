// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode image with bitDepth of 1 encode a 1x5 image 1`] = `
Uint8Array [
  66,
  77,
  166,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  146,
  0,
  0,
  0,
  124,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  5,
  0,
  0,
  0,
  1,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  0,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  0,
  66,
  71,
  82,
  115,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  4,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  255,
  255,
  255,
  0,
  128,
  0,
  0,
  0,
  128,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  128,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode a 5x1 image 1`] = `
Uint8Array [
  160,
  0,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode a 5x5 image 1`] = `
Uint8Array [
  66,
  77,
  166,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  146,
  0,
  0,
  0,
  124,
  0,
  0,
  0,
  5,
  0,
  0,
  0,
  5,
  0,
  0,
  0,
  1,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  20,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  0,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  255,
  0,
  0,
  0,
  66,
  71,
  82,
  115,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  4,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  255,
  255,
  255,
  0,
  0,
  0,
  0,
  0,
  56,
  0,
  0,
  0,
  40,
  0,
  0,
  0,
  56,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode a 6x4 image 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  252,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  252,
  0,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode a 10x2 image 1`] = `
Uint8Array [
  169,
  192,
  0,
  0,
  229,
  64,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode a 62x4 1`] = `
Uint8Array [
  128,
  0,
  0,
  127,
  255,
  128,
  0,
  0,
  255,
  255,
  255,
  0,
  0,
  255,
  255,
  252,
  0,
  0,
  0,
  255,
  255,
  0,
  0,
  0,
  255,
  255,
  255,
  0,
  0,
  255,
  255,
  252,
]
`;

exports[`encode image with bitDepth of 1 encode image where skipBit can equal relOffset on the last column 1`] = `
Uint8Array [
  15,
  255,
  255,
  255,
  255,
  255,
  255,
  240,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  15,
  255,
  255,
  255,
  255,
  255,
  255,
  240,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`encode image with bitDepth of 1 encode image with exactly 4 bytes width 1`] = `
Uint8Array [
  255,
  255,
  0,
  0,
  0,
  0,
  255,
  255,
]
`;

exports[`encode image with bitDepth of 1 encode image with more that 4 bytes width 1`] = `
Uint8Array [
  0,
  0,
  255,
  255,
  255,
  192,
  0,
  0,
  0,
  0,
  255,
  255,
  255,
  192,
  0,
  0,
]
`;
